Yoast/duplicate-post
# Refactor and update Duplicate Post options

## Context


* Refactors the old Duplicate Post options page to use a newer, OOP approach.
* Introduces a new format for storing similar types of options (`duplicate_post_show_link` and `duplicate_post_show_link_in`) and migrates the old options for `duplicate_post_show_link`.

## Summary


This PR can be summarized in the following changelog entry:

* Introduces new settings to individually enable/disable the `New Draft`, `Clone` and `Rewrite & Republish` links.

## Relevant technical choices:

* This PR introduces a large overhaul of the way the options were registered and how the forms were being built, as to reduce the amount of duplication that was previously present.
* Adds tests for (nearly) all changed code.

## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

#### Test migration of settings
* Before checking out this branch, please build `master` and ensure you've altered some of the settings for the 'Show links in' section, which you can find in Settings -> Duplicate Post -> Display tab. Don't forget to save your changes.
* Verify that these changes you've applied, are propagated in the UI (i.e. New Draft, Clone etc. no longer show up in the Posts overview, for example).
* Checkout this branch, build it (run both `composer install` and `grunt build`) and refresh the page.
* See that the changes you made in the settings page, are still the same for the 'Show links in' section. Please also note the addition of the 'Show these links' section.
* Verify that the changes you made in the second step, still apply to the UI to ensure settings have properly been migrated.

#### Test toggling individual links
* In the Settings -> Duplicate Post -> Display tab, change one or more of the options in the 'Show these links' section and save them.
* Navigate to the Posts overview and verify that the links you have disabled, don't show in the Bulk actions drop-down, nor do they show in the Posts overview, when hovering over a post.
* In case you've disabled 'New Draft', create a new post. See that the 'Copy to a new draft' link is not present in the Publish box (both in Classic and in Block Editor). Also verify that the link is no longer present in the Posts overview, the admin bar and Bulk actions drop-down.
* In case you've disabled 'Rewrite & Republish', open an existing post. See that the 'Rewrite & Republish' link is not present in the Publish box (both in Classic and in Block Editor). Also verify that the link is no longer present in the Posts overview, the admin bar and Bulk actions drop-down.
* In case you've disabled 'Clone', go to the posts overview. See that the 'Clone' link is not present in the Posts overview or the Bulk actions drop-down.

#### Testing the other options

**Although these options haven't changed in how they function, its still a good idea to check these for any regressions**

* Navigate to Settings -> Duplicate Post and edit any of the other settings. Save your changes and verify that the settings are executed as expected.

### Test instructions for QA when the code is in the RC




QA can test this PR by following these steps:

#### Test migration of settings
* Before installing the RC, please install the current stable version and ensure you've altered some of the settings for the 'Show links in' section, which you can find in Settings -> Duplicate Post -> Display tab. Don't forget to save your changes.
* Verify that these changes you've applied, are propagated in the UI (i.e. New Draft, Clone etc. no longer show up in the Posts overview, for example).
* Install the RC and navigate bad to Settings -> Duplicate Post -> Display tab.
* See that the changes you made in the settings page, are still the same for the 'Show links in' section. Please also note the addition of the 'Show these links' section.
* Verify that the changes you made in the second step, still apply to the UI to ensure settings have properly been migrated.

#### Test toggling individual links
* In the Settings -> Duplicate Post -> Display tab, change one or more of the options in the 'Show these links' section and save them.
* Navigate to the Posts overview and verify that the links you have disabled, don't show in the Bulk actions drop-down, nor do they show in the Posts overview, when hovering over a post.
* In case you've disabled 'New Draft', create a new post. See that the 'Copy to a new draft' link is not present in the Publish box (both in Classic and in Block Editor). Also verify that the link is no longer present in the Posts overview, the admin bar and Bulk actions drop-down.
* In case you've disabled 'Rewrite & Republish', open an existing post. See that the 'Rewrite & Republish' link is not present in the Publish box (both in Classic and in Block Editor). Also verify that the link is no longer present in the Posts overview, the admin bar and Bulk actions drop-down.
* In case you've disabled 'Clone', go to the posts overview. See that the 'Clone' link is not present in the Posts overview or the Bulk actions drop-down.

#### Testing the other options

**Although these options haven't changed in how they function, its still a good idea to check these for any regressions**

* Navigate to Settings -> Duplicate Post and edit any of the other settings. Save your changes and verify that the settings are executed as expected.

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

* All options have been migrated to use a new approach regarding being built up for use in the UI. It's a good idea to ensure that these still save and execute as expected.

## UI changes


## Documentation


## Quality assurance


Fixes [P3-250]
